### PR TITLE
fix: remove some typing errors

### LINKED ISSUE #191: Fix typing errros, other warnings

### PR DESCRIPTION
**Pull request by David-Kristek**
---
## Status
**READY**

## Which issue(s) does this pull-request fix?
Closes #191 

## Any special notes for your reviewer?

---

## Checklist
- [ ] Automated tests exist
- [ ] Updated Package Requirements (if required, and with maintainers' approval)
- [ ] Local unit tests performed
- [ ] Documentation exists [link]()
- [ ] Local pre-commit hooks performed
- [ ] Desired commit message set as PR title and description set above
- [ ] Link to relevant GitHub issue provided

### CODE REVIEW

File: src/genai/schemas/descriptions.py [hunk 1]
```diff
@@ -73,7 +73,6 @@ class TunesAPIDescriptions:
     VERBALIZER = "Verbalizer template to be used for formatting data at train and inference time. This template may use double brackets to indicate where fields from training data should be rendered. The template can contain one or both of {{input}} and {{output}}. Defaults to {{input}}."
     OFFSET = "Specifies the starting position in the collection. Defaults to 0."
     LIMIT = "Specifies the maximum number of items in the collection that should be returned. Defaults to 100. Maximum is 100."
-    SEARCH = "Filters the items to be returned based on their name."
```
Comment (David-Kristek) [inline src/genai/schemas/descriptions.py:76]: Duplicate

File: src/genai/schemas/responses.py [hunk 2]
```diff
@@ -256,53 +256,6 @@ class TuneMethodsGetResponse(GenAiResponseModel):
...
-    learning_rate: Optional[float]
-    max_input_tokens: Optional[int]
-    max_output_tokens: Optional[int]
-    num_epochs: Optional[int]
-    num_virtual_tokens: Optional[int]
-    verbalizer: Optional[str]
-
-
-class TuneInfoResult(GenAiResponseModel):
-    id: str
```
Comment (David-Kristek) [inline src/genai/schemas/responses.py:286]: Duplicates

File: src/genai/extensions/pandas/prompt_sub.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ def sub_from_dataframe(
         strategy: Literal["random", "sequential", "sample"] = "sequential",
         start_index: int = 0,
         n: int = 1,  # prompt number to be returned - Optional
-    ) -> Union["PromptPattern", list["PromptPattern"]]:
+    ) -> list["PromptPattern"]:
```
Comment (Tomas2D) [inline src/genai/extensions/pandas/prompt_sub.py:28]: This is probably not correct based on the implementation. Can you try it out?

File: src/genai/model.py [hunk 4]
```diff
@@ -184,7 +183,7 @@ def generate_async(
                 to be called after generating result for a prompt.
             hide_progressbar (bool, optional): boolean flag to hide or show a progress bar.
             options (Options, optional): Additional parameters to pass in the query payload. Defaults to None.
-
+            throw_on_error(bool, optional): Throws error on failure.
```
Comment (Tomas2D) [inline src/genai/model.py:186]: You are missing a default value.

File: src/genai/services/tune_manager.py [hunk 5]
```diff
@@ -56,7 +57,9 @@ def list_tunes(
             raise GenAiException(e)
 
     @staticmethod
-    def get_tune(tune_id: str, credentials: Credentials = None, service: ServiceInterface = None) -> TuneGetResponse:
+    def get_tune(
+        tune_id: str, credentials: Credentials = None, service: ServiceInterface = None
+    ) -> Union[TuneInfoResult, None]:
```
Comment (Tomas2D) [inline src/genai/services/tune_manager.py:62]: `Union[TuneInfoResult, None]` -> `Optional[TuneInfoResult]`

**Review by Tomas2D — CHANGES_REQUESTED** [review]
Great work so far!

Besides these small changes, I saw some try/catch blocks with non-explicit exceptions to catch.

```
        try:
            response = service._prompt_templating.update_prompt_templates(id=id, name=name, value=template)
            if response.status_code == 200:
                response_result = WatsonxTemplate(**response.json()["results"])
                return response_result
            raise GenAiException(response)
        except Exception as ex:
            raise GenAiException(ex)
```

This is firstly bad practice, but for now, it is sufficient to handle cases like this in the following way:

```
        except Exception as ex:
            raise to_genai_error(ex)
```

File: src/genai/prompt_pattern.py [hunk 6]
```diff
@@ -510,7 +506,7 @@ def _sub_from_tabular_data(
 
             start_index += len(variables)
 
-        return complete_prompt
```
Comment (David-Kristek) [inline src/genai/prompt_pattern.py:517]: This didn't return the correct type - only arrays.

File: src/genai/services/async_generator.py [hunk 7]
```diff
@@ -194,7 +194,7 @@ def _request_launcher(self):
         except Exception as ex:
             self.throw_on_error = True
             self._queue.put_nowait((1, 1, None, ex))
-            raise ex
+            raise to_genai_error(ex)
```
Comment (Tomas2D) [inline src/genai/services/async_generator.py:197]: Could you keep it as is? We want the same error in the queue (line above) to be the same as the error we just raised.

File: tests/test_prompt_pattern.py [hunk 8]
```diff
@@ -158,6 +158,7 @@ def test_sub_from_csv_random(self):
             },
         )
 
+        assert isinstance(res, PromptPattern) is True
```
Comment (Tomas2D) [inline tests/test_prompt_pattern.py:161]: 👍🏻

**Review by Tomas2D — CHANGES_REQUESTED** [review]

**Review by Tomas2D — APPROVED** [review]

**Pull request merged**